platforms/core-execution/persistent-cache/src/testFixtures/groovy/org/gradle/cache/internal/TestCrossBuildInMemoryCacheFactory.groovy

```diff
@@ -58,12 +60,17 @@ class TestCrossBuildInMemoryCacheFactory implements CrossBuildInMemoryCacheFacto
 
         @Override
         V get(K key, Function<? super K, ? extends V> factory) {
-            def v = values.get(key)
-            if (v == null) {
-                v = factory.apply(key)
-                values.put(key, v)
+            try {
+                lock.lock()
```

Comment:
💭 This test cache could call `factory` in thread-unsafe way which caused issues in some tests after changes. For example we generated the same Decorated class multiple times and then ClassLoader threw an exception that the same class was already define (LinkageError).

Note: using `values.computeIfAbsent()` doesn't work here, since we can get `recursive update exception`.